gtest/src/manager.rs

```diff
@@ -1,97 +1,249 @@
+
+                    let (bytes, exit_code) = match reply_payload {
+                        Ok(payload) => {
+                            if let DispatchKind::Init = kind {
+                                self.init_success(message.id(), message.dest());
+                            }
+
+                            (payload, 0)
+                        }
+                        Err(trap) => {
+                            if let DispatchKind::Init = kind {
+                                self.message_dispatched(DispatchOutcome::InitFailure {
+                                    message_id: message.id(),
+                                    program_id: message.dest(),
+                                    origin: message.source(),
+                                    reason: "Err(_) returned in function",
+                                })
+                            } else {
+                                self.message_dispatched(DispatchOutcome::MessageTrap {
+                                    message_id: message.id(),
+                                    program_id: message.dest(),
+                                    trap: Some("Err(_) returned in function"),
+                                })
+                            }
+
+                            (trap.as_bytes().to_vec(), 1)
+                        }
+                    };
+
+                    let nonce = self.fetch_inc_message_nonce();
+
+                    if exit_code != 0 || !bytes.is_empty() {
+                        self.send_message(
+                            message.id(),
+                            Message::new_reply(
+                                nonce.into(),
+                                message.dest(),
+                                message.source(),
+                                bytes.into(),
+                                message.gas_limit(),
+                                0,
+                                message.id(),
+                                exit_code,
+                            ),
+                        );
+                    }
+                }
+            }
         }
     }
-}
 
-impl CollectState for InMemoryExtManager {
-    fn collect(&self) -> State {
-        let InMemoryExtManager {
-            message_queue,
-            log,
-            programs,
-            current_failed,
-            ..
-        } = self.clone();
-
-        State {
-            message_queue,
-            log,
-            programs: programs.into_inner(),
-            current_failed,
+    fn clear(&mut self) {
+        self.log.clear();
+        self.failed = false;
+
+        if !self.message_queue.is_empty() {
+            panic!("Message queue isn't empty");
         }
     }
+
+    fn init_success(&mut self, message_id: MessageId, program_id: ProgramId) {
+        self.programs
+            .get_mut(&program_id)
+            .expect("Can't find existing program")
+            .1 = ProgramState::Initialized;
```

Comment:
Again for readability:
```
let (_, ref mut state, ..) = self.programs.get_mut(&program_id).expect(/* ... */);
state = ProgramState::Initialized;
```

Reply:
Done with the same idea and a bit different syntax